Update Fetch-Canary-Syslog.ps1

Xorlent opened the pull request:
## Change description
Powershell syslog fetcher (Fetch-Canary-Syslog.ps1): Fixed single-line parse issue with simplified string -split regex.
Cloudlfare workers (canary-receiver.js): Now filtering IP addresses added to blocklist KV store based on event description.

## Types of changes

What types of changes does your code introduce to this repository?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] Lint and unit tests pass locally with my changes (if applicable)
- [ ] I have run pre-commit (`pre-commit` in the repo)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Linked to the relevant github issue or github discussion

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

mclmax commented:
Hi there,

Thanks for sending the PR. I've taken a quick look. Comments to follow on the code.

Max

mclmax commented on Cloudflare/Fetch-Canary-Syslog.ps1:
```diff
@@ -16,7 +16,7 @@ $Response = Invoke-WebRequest -Uri $WorkerURL -Headers @{ 'auth' = $WorkerAuth}
  
 $Content = $Response.Content
-$SyslogArray = ($Content -split "\r?\n|\r")
+$SyslogArray = ($Content -split '\r?\n')
```
Could you explain why this is necessary here? Perhaps with an example of what you're fixing / changing? Thanks

Xorlent replied:
Hi Max, for instances where there is only a single returned value, this was causing a blank value.  This resolved the issue for my test, but I may end up reworking the code in the long run.

Xorlent replied:
This was rewritten

mclmax commented on Cloudflare/canary-receiver.js:
```diff
@@ -71,12 +71,29 @@ async function handleRequest(request) {
-        // process  
-        // Store in KV store (Key-Value store)
-        await canaryblocks.put(MaliciousIP, Timestamp, {expirationTtl: IPBlocklistTTL})
+        // process
+        switch (true){
+          case /Scan/.test(data.Description):
+            // Store in KV store (Key-Value store)
+            await canaryblocks.put(MaliciousIP, Timestamp, {expirationTtl: IPBlocklistTTL})
+          break;
+          case /Load/.test(data.Description):
+            break;
+          case /Settings Changed/.test(data.Description):
+            break;
+          case /Disconnected/.test(data.Description):
+            break;
+          case /Reconnected/.test(data.Description):
+            break;
+          case /Dummy/.test(data.Description):
+            break;
+          default:
+            // Store in KV store (Key-Value store)
+            await canaryblocks.put(MaliciousIP, Timestamp, {expirationTtl: IPBlocklistTTL})
+          break;
+        } // end switch
```
If i'm following here you'd like to exclude these IPs from the block list?

I think that's reasonable. Just a few things:

- /Load/ is going to match against legit alerts, eg. our HTTP Page Load incident will get dropped here. So we'll have to remove that. The others look fine though.
- /Scan/ just looks like the default case in the switch so i think we can clean this up a little bit.
- I think we can do this without a switch statement. It's a little difficult to follow the flow here.
- We're breaking out of the switch and then replying with the ``Marked IP: ${MaliciousIP} and Stored Event: ${Payload}`,` response which isn't happening, right?

Let's rework this slightly to into an if statement with a function:

```
const isHouseKeepingAlert = () => {
  return Settings Changed/.test(data.Description) || /Disconnected/.test(data.Description) || /Reconnected/.test(data.Description) || /Dummy/.test(data.Description);
};

if (isHouseKeepingAlert()) {
  // return response here with the message something like... 'Not marking house keeping alert but storing event for syslog'
}

await canaryblocks.put(MaliciousIP, Timestamp, {expirationTtl: IPBlocklistTTL})

etc...
```
Let me know what you think?

Xorlent replied:
Good timing, and I was making some adjustments to our deployed workers and removed the first switch case, since it was redundant.  I like your version much better.  Thank you.

Xorlent reviewed (commented):
Fetch-Canary-Syslog.ps1: I found an issue with the PowerShell Invoke-WebRequest if mshtml (IE) is disabled.  This version fixes this.
canary-receiver.js: Refactored based on feedback, and also account for an event type "Custom Service."   If the Thinkst team have test automation, I'd love to get output from it of all the different event types to make sure I am covering every case!

mclmax commented:
Thanks @Xorlent!  Sorry for the delayed response here. I've let some small comments on the code. It's looking great! Just one nitpick and then I think we are there :) Looking forward to have this included!

Xorlent commented:
I'm sorry Max, I don't see the comment you're referring to, unless you mean the PowerShell regex comment from a few weeks ago.  I rewrote that entire section of code at the same time I addressed an issue if MSHTML is disabled on the host.

mclmax commented:
Meh, sorry @Xorlent. Forgot to submit the comments. Github is hard

Xorlent commented:
Okay, all done.  Thanks Max.

mclmax commented:
Thanks for contributing @Xorlent . We really appreciate folks taking the time. Please email me: max at thinkst dot com for a little something to say thanks!
